hs/CalendarResponse.hs

```diff
@@ -0,0 +1,281 @@
+zip' :: [a] -> [b] -> [c] -> [(a,b,c)]
+zip' (x:xs) (y2:ys) (z:zs) = (x,y2,z):zip' xs ys zs
+zip' _ _ _ = []
+
+-- | Creates an event for each start/end time.
+eventsByDate :: String -- ^ Course code.
+             -> String -- ^ Course section.
+             -> String -- ^ Current time.
+             -> ([String], -- ^ Start time
+             [String], -- ^ End time
+             (String, String)) -- ^ Start/End date.
+             -> [String]
+eventsByDate code sect timeSystem (start, end, dates) =
+    (zip start end) >>= (eventsGenerator code sect timeSystem dates)
+
+-- | Generates the string that represents each event.
+eventsGenerator :: String -- ^ Course code.
+                -> String -- ^ Course section.
+                -> String -- ^ Current time.
+                -> (String, String) -- ^ Start/End date.
+                -> (String, String) -- ^ Start/End time.
+                -> [String]
+eventsGenerator code sect timeSystem (startDate, endDate) (start, end) =
+    if startDate == "" || endDate == ""
+    then []
+    else ["BEGIN:VEVENT", "DTSTAMP:" ++ timeSystem,
+          "DTSTART;TZID=America/Toronto:" ++ startDate ++ start,
+          "DTEND;TZID=America/Toronto:" ++ startDate ++ end,
+          "RRULE:FREQ=WEEKLY;UNTIL=" ++ endDate ++ "000000Z",
+          "ORGANIZER:University of Toronto",
+          "SUMMARY:" ++ code ++ " " ++ sect,
+          "CATEGORIES:EDUCATION", "END:VEVENT"]
+
+-- ** Ordering data
+
+-- | Orders the time fields for each course; which were obtained from the
+-- database
+orderData :: [Time] -> [[[Double]]]
+orderData courseFields = filter notNull (assignDay $ map timeField courseFields)
+    where
+        notNull field = not $ null field
+
+-- | Organizes the data by the day of the week, assigning a position based on
+-- the day in the time field.
+assignDay :: [[Double]] -> [[[Double]]]
+assignDay lst = [monday, tuesday, wednesday, thursday, friday]
+    where
+        monday = filter (\ field -> field !! 0 == 0.0) lst
+        tuesday = filter (\ field -> field !! 0 == 1.0) lst
+        wednesday = filter (\field -> field !! 0 == 2.0) lst
+        thursday = filter (\field -> field !! 0 == 3.0) lst
+        friday = filter (\field -> field !! 0 == 4.0) lst
+
+-- ** Start time
+
+-- | Obtains a list of start times for each course.
+startTime :: [[[Double]]] -> [[String]]
+startTime dataInOrder = map checkTimeStart dataInOrder
+    where
+        checkTimeStart dataByDay = 
+            map getStrTime ([head (sortList dataByDay)] ++
+                        getStartConsecutives (sortList dataByDay))
+        sortList dataDay = sort $ map (!! 1) dataDay
+
+-- | Gets the start times that are not the very first start time. 
+getStartConsecutives :: [Double] -> [Double]
+getStartConsecutives lst = 
+    filter (/= 30.0) ([if lst !! i == lst !! (i + 1) - 0.5
+                       then 30.0
+                       else lst !! (i + 1)| i <- [0 .. l]])
+    where
+        l = length lst - 2
```

Comment:
What is the significance of `length lst - 2`? A more appropriate name would drastically increase readability.